File: vscode-extension/src/diagramExporter.ts
import * as vscode from "vscode"
import { diagramTitle, exportTargetPath } from "./diagramDocument"
import type {
  ExportFormat,
  HostMessage,
  WebviewMessage,
} from "./shared/protocol"

/** A canvas that never answers should not hold a write open forever. */
const RENDER_TIMEOUT_MS = 30_000

/** The editor closed before it could render. Not a failure worth reporting. */
class ExportCancelled extends Error {}

type ExportOutcome = Extract<
  WebviewMessage,
  { type: "exportResult" } | { type: "exportFailed" }
>

/**
 * Renders diagrams to sibling image files.
 *
 * Only the canvas can render a diagram, so every export is a round trip: the
 * host asks a webview, the webview answers, and the reply is matched back to its
 * request by id. One exporter serves every open editor, since a request outlives
 * neither more nor less than the webview it was sent to.
 */
export class DiagramExporter {
  private sequence = 0
  private readonly inFlight = new Map<
    number,
    { webview: vscode.Webview; settle: (result: string | Error) => void }
  >()

  /** Render the diagram in `webview` and write the image next to `uri`. */
  public async write(
    uri: vscode.Uri,
    webview: vscode.Webview,
    format: ExportFormat,
    options: { silent: boolean }
  ): Promise<void> {
    // Both the command and auto-export come through here, so the trust promise
    // the manifest makes — "writing sibling image exports is disabled until you
    // trust the workspace" — is kept on each.
    if (!vscode.workspace.isTrusted) {
      if (!options.silent) {
        void vscode.window.showWarningMessage(
          "Trust this workspace to export diagram images."
        )
      }
      return
    }
    const name = `${diagramTitle(uri.path)}.${format}`
    try {
      const payload = await this.render(webview, format)
      const bytes =
        format === "png"
          ? Buffer.from(payload, "base64")
          : Buffer.from(payload, "utf8")
      const target = uri.with({ path: exportTargetPath(uri.path, format) })
      await vscode.workspace.fs.writeFile(target, bytes)
      if (options.silent) {
        vscode.window.setStatusBarMessage(`Apollon: exported ${name}`, 2000)
      } else {
        void vscode.window.showInformationMessage(`Exported ${name}`)
      }
    } catch (error) {
      if (error instanceof ExportCancelled) {
        return
      }
      void vscode.window.showErrorMessage(
        `Apollon could not export ${name}: ${error instanceof Error ? error.message : String(error)}`
      )
    }
  }

  /** Hand a webview's answer back to the request that is waiting for it. */
  public settle(message: ExportOutcome): void {
    const entry = this.inFlight.get(message.requestId)
    entry?.settle(
      message.type === "exportFailed"
        ? new Error(message.reason)
        : message.payload
    )
  }

  /** A closed editor will never answer; settle its exports now, not in 30s. */
  public cancel(webview: vscode.Webview): void {
    for (const entry of [...this.inFlight.values()]) {
      if (entry.webview === webview) {
        entry.settle(new ExportCancelled())
      }
    }
  }

  private render(
    webview: vscode.Webview,
    format: ExportFormat
  ): Promise<string> {
    const requestId = ++this.sequence
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.inFlight.delete(requestId)
        reject(new Error("the editor did not respond"))
      }, RENDER_TIMEOUT_MS)
      this.inFlight.set(requestId, {
        webview,
        settle: (result) => {
          clearTimeout(timeout)
          this.inFlight.delete(requestId)
          if (result instanceof Error) {
            reject(result)
          } else {
            resolve(result)
          }
        },
      })
      void webview.postMessage({
        type: "export",
        format,
        requestId,
      } satisfies HostMessage)
    })
  }
}

File: jetbrains-plugin/src/main/kotlin/de/tum/cit/aet/apollon/export/DiagramExporter.kt
package de.tum.cit.aet.apollon.export

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.vfs.LocalFileSystem
import de.tum.cit.aet.apollon.document.diagramTitle
import de.tum.cit.aet.apollon.document.exportTargetPath
import de.tum.cit.aet.apollon.protocol.ExportFormat
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.Base64
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

private const val RENDER_TIMEOUT_SECONDS = 30L

/**
 * Renders diagrams to sibling image files.
 *
 * Ported from the VS Code extension's `DiagramExporter`: only the canvas can
 * render a diagram, so every export is a round trip through the JS bridge —
 * the host sends an `export` [HostMessage], the webview answers with an
 * `exportResult`/`exportFailed` [WebviewMessage] matched back by [settle]'s
 * `requestId`. One exporter serves every open editor.
 */
class DiagramExporter {
    private val sequence = AtomicInteger()
    private val inFlight = ConcurrentHashMap<Int, CompletableFuture<String>>()

    /**
     * Starts a render/write round trip. [sendExportRequest] must post the
     * `export` message to the webview with the given `requestId`; the result
     * (or timeout) arrives later through [settle].
     */
    fun write(
        filePath: String,
        format: ExportFormat,
        silent: Boolean,
        sendExportRequest: (requestId: Int) -> Unit,
    ) {
        val requestId = sequence.incrementAndGet()
        val future = CompletableFuture<String>()
        inFlight[requestId] = future
        sendExportRequest(requestId)

        val name = "${diagramTitle(filePath)}.${format.name}"
        future
            .orTimeout(RENDER_TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .whenComplete { payload, error ->
                inFlight.remove(requestId)
                if (error == null) {
                    writeSibling(filePath, format, payload, name, silent)
                } else if (error !is CancellationException) {
                    reportFailure(name, error.message ?: error.toString())
                }
            }
    }

    /** Hand a webview's answer back to the request that is waiting for it. */
    fun settle(
        requestId: Int,
        payload: String?,
        reason: String?,
    ) {
        val future = inFlight[requestId] ?: return
        if (payload != null) {
            future.complete(payload)
        } else {
            future.completeExceptionally(RuntimeException(reason ?: "export failed"))
        }
    }

    /** A closed editor will never answer; cancel its exports now, not in 30s. */
    fun cancelAll() {
        inFlight.values.forEach { it.cancel(true) }
        inFlight.clear()
    }

    private fun writeSibling(
        sourcePath: String,
        format: ExportFormat,
        payload: String,
        name: String,
        silent: Boolean,
    ) {
        try {
            val bytes =
                if (format == ExportFormat.png) {
                    Base64.getDecoder().decode(payload)
                } else {
                    payload.toByteArray(Charsets.UTF_8)
                }
            val target = Path.of(exportTargetPath(sourcePath, format.name))
            Files.write(target, bytes)
            LocalFileSystem.getInstance().refreshAndFindFileByNioFile(target)
            if (!silent) {
                ApplicationManager.getApplication().invokeLater {
                    Messages.showInfoMessage("Exported $name", "Apollon")
                }
            }
        } catch (e: IOException) {
            reportFailure(name, e.message ?: e.toString())
        } catch (e: IllegalArgumentException) {
            // Malformed base64 payload (Base64.getDecoder().decode) or an
            // invalid sibling path (Path.of) both surface as this.
            reportFailure(name, e.message ?: e.toString())
        }
    }

    private fun reportFailure(
        name: String,
        reason: String,
    ) {
        ApplicationManager.getApplication().invokeLater {
            Messages.showErrorDialog("Apollon could not export $name: $reason", "Export Failed")
        }
    }
}
